File: frontend/src/api/checklist.ts
import { request } from './client'

export type ChecklistResult = 'not_assessed' | 'satisfactory' | 'unsatisfactory' | 'not_applicable'
export type ChecklistStatus = 'draft' | 'submitted'
export type ClarificationStatus = 'none' | 'open' | 'answered' | 'resolved' | 'withdrawn'

export interface ChecklistItemDef {
  key: string
  section: string
  title: string
  guidance: string
  applicable_by_default: boolean
}

export interface ChecklistSection {
  key: string
  title: string
  items: ChecklistItemDef[]
}

export interface ChecklistSchema {
  version: number
  description: string
  sections: ChecklistSection[]
  item_count: number
}

export interface ChecklistItem {
  id: string
  key: string
  section: string
  title: string
  guidance: string
  position: number
  result: ChecklistResult
  comment: string | null
  needs_clarification: boolean
  clarification_status: ClarificationStatus
}

export interface ChecklistCounts {
  total: number
  assessed: number
  flagged: number
  unsatisfactory: number
  not_applicable: number
  /** Unsatisfactory or flagged items without a comment: what still blocks a submit. */
  missing_comments: number
}

export interface Checklist {
  id: string
  application_id: string
  visit_no: number
  schema_version: number
  status: ChecklistStatus
  version: number
  created_by: string
  created_at: string
  updated_at: string | null
  submitted_by: string | null
  submitted_at: string | null
  counts: ChecklistCounts
  /** What still blocks a submit, in a sentence; null when nothing does. */
  remaining: string | null
  items: ChecklistItem[]
}

/** The line the case rail shows (S-31 summary). */
export interface ChecklistSummary {
  visit_no: number
  status: ChecklistStatus
  version: number
  counts: ChecklistCounts
  updated_at: string | null
  submitted_at: string | null
}

export interface ChecklistItemInput {
  key: string
  result: ChecklistResult
  comment: string | null
  needs_clarification: boolean
}

export interface ChecklistSaveInput {
  items: ChecklistItemInput[]
  version: number
  /** Client-generated per save; a replayed one answers with the current state. */
  save_id?: string
}

export function getChecklistSchema(): Promise<ChecklistSchema> {
  return request<ChecklistSchema>('/checklist-schema')
}

/** Creates the current visit's draft on first open, returns it afterwards. */
export function openChecklist(id: string): Promise<Checklist> {
  return request<Checklist>(`/officer/applications/${id}/checklist`, { method: 'POST' })
}

export function getChecklist(id: string, visit?: number): Promise<Checklist> {
  return request<Checklist>(`/officer/applications/${id}/checklist${visit ? `?visit=${visit}` : ''}`)
}

export function saveChecklist(id: string, body: ChecklistSaveInput): Promise<Checklist> {
  return request<Checklist>(`/officer/applications/${id}/checklist`, { method: 'PUT', body })
}

File: backend/app/schemas/checklist.py
"""The site visit checklist (US-060 to US-063) as the officer works on it and as the case summarises
it. Operators never receive these models (US-064 serves them a separate clarification view)."""

import uuid
from datetime import datetime

from pydantic import BaseModel, Field


class ChecklistItemDefOut(BaseModel):
    key: str
    section: str
    title: str
    guidance: str
    applicable_by_default: bool


class ChecklistSectionOut(BaseModel):
    key: str
    title: str
    items: list[ChecklistItemDefOut]


class ChecklistSchemaOut(BaseModel):
    version: int
    description: str
    sections: list[ChecklistSectionOut]
    item_count: int


class ChecklistItemOut(BaseModel):
    id: uuid.UUID
    key: str
    section: str
    title: str
    guidance: str
    position: int
    result: str
    comment: str | None
    needs_clarification: bool
    clarification_status: str
    # Extra findings (US-092): the officer's own title, and the template item it sits under when linked.
    is_extra: bool = False
    custom_title: str | None = None
    parent_key: str | None = None


class ChecklistCounts(BaseModel):
    total: int
    assessed: int
    flagged: int
    unsatisfactory: int
    not_applicable: int
    # Unsatisfactory or flagged items without a comment: what still blocks a submit.
    missing_comments: int


class ChecklistOut(BaseModel):
    id: uuid.UUID
    application_id: uuid.UUID
    visit_no: int
    schema_version: int
    status: str
    version: int
    created_by: str
    created_at: datetime
    updated_at: datetime | None
    submitted_by: str | None
    submitted_at: datetime | None
    counts: ChecklistCounts
    # What stands between the draft and a submit, in a sentence; None when nothing does.
    remaining: str | None
    items: list[ChecklistItemOut]


class ChecklistSummaryOut(BaseModel):
    """The line the case rail shows (S-31 summary card)."""

    visit_no: int
    status: str
    version: int
    counts: ChecklistCounts
    updated_at: datetime | None
    submitted_at: datetime | None


class ChecklistItemIn(BaseModel):
    # None for a new extra finding: the server assigns its key and returns it (US-092).
    key: str | None = None
    result: str
    comment: str | None = Field(default=None, max_length=2000)
    needs_clarification: bool = False
    custom_title: str | None = Field(default=None, max_length=120)
    parent_key: str | None = Field(default=None, max_length=48)


class ChecklistSaveIn(BaseModel):
    items: list[ChecklistItemIn]
    version: int
    # Client-generated per save; a replayed one answers with the current state (US-061).
    save_id: str | None = Field(default=None, max_length=64)


__all__ = [
    "ChecklistCounts",
    "ChecklistItemDefOut",
    "ChecklistItemIn",
    "ChecklistItemOut",
    "ChecklistOut",
    "ChecklistSaveIn",
    "ChecklistSchemaOut",
    "ChecklistSectionOut",
    "ChecklistSummaryOut",
]
